### PR TITLE
Feature/400 enhancejs and cookie logic on server side

### PR DESCRIPTION
**Pull request by radibit**

### CODE REVIEW

**Comment by stefanjudis** [comment]
oh man you made it!!! :tada: 

will have a look at it tomorrow. :+1: :+1: :+1:

**Comment by radibit** [comment]
and it's working like a charm :wink:

**Comment by stefanjudis** [comment]
One thing that pops to my mind first - can wew squash this commits together. Usually I don't think it's necessary if there are a couple of commits. But you've got a ton?

What do you think?

File: app.js [hunk 1]
```diff
@@ -9,7 +9,8 @@ var minify      = require( 'html-minifier' ).minify;
 var request     = require( 'request' );
 var config      = require( './config/config' );
 var async       = require( 'async' );
-
+var cookieParser = require( 'cookie-parser' );
+var revisions   = require( './public/rev');
```
Comment (stefanjudis) [inline app.js]: Coding style. :blush:

File: app.js [hunk 2]
```diff
@@ -53,9 +54,10 @@ var pages = {
  */
 var pageContent = {
   css       : fs.readFileSync( './public/main.css', 'utf8' ),
+  enhance   : fs.readFileSync( './public/enhance.js', 'utf8' ),
   hashes    : {
-    css : md5( fs.readFileSync( './public/main.css', 'utf8' ) ),
-    js  : md5( fs.readFileSync( './public/tooling.js', 'utf8' ) )
+    css     : revisions.styles,
+    js      : md5( fs.readFileSync( './public/tooling.js', 'utf8' ) )
   },
```
Comment (stefanjudis) [inline app.js]: Can we change the structure of scripts also for consistency.

File: gulpfile.js [hunk 3]
```diff
@@ -83,10 +94,19 @@ gulp.task( 'styles', function () {
...
+    .pipe( gulp.dest( 'public/' ) )
+    .pipe( gutil.buffer( function ( err, dataFiles ) {
+      gulp.src( files.rev )
+        .pipe( jsoneditor( {
+          'styles': dataFiles.map( function ( dataFile ) {
+            return dataFile.path.replace( dataFile.base, '' ).slice( 0, -4 ).split( '-' )[ 1 ];
+          } ).join( '' )
+        } ) )
+        .pipe( gulp.dest( 'public/' ) );
+    } ) );
```
Comment (stefanjudis) [inline gulpfile.js]: I'm not quite sure about this one, but I think the stream might be returned be the function in `gutil.buffer`.

File: package.json [hunk 4]
```diff
@@ -18,26 +18,34 @@
...
     "gulp-jshint": "^1.9.0",
+    "gulp-json-editor": "^2.2.1",
     "gulp-jsonlint": "^1.0.2",
     "gulp-less": "3.0.0",
     "gulp-minify-css": "~0.3.0",
+    "gulp-rev": "^3.0.1",
     "gulp-svgstore": "~4.0.1",
     "gulp-task-listing": "^1.0.0",
+    "gulp-template": "^3.0.0",
     "gulp-uglify": "^1.0.1",
```
Comment (stefanjudis) [inline package.json]: Leftover?

File: templates/partials/head.tpl [hunk 5]
```diff
@@ -29,11 +29,19 @@
...
   </script>
 
+  <meta name="maincss"  content="<%= cdn %>/main-<%= hash.css %>.css">
+  <% if ( !cssCookie || cssCookie !== hash.css ) { %>
+  <style><%= css %></style>
+  <% } else { %>
+  <link rel="stylesheet" href="<%= cdn %>/main-<%= hash.css %>.css">
+  <% } %>
+  <script><%= enhance %></script>
+
```
Comment (stefanjudis) [inline templates/partials/head.tpl]: We also need a `noscript` version, or?

File: package.json (hunk 4 above)
Comment (radibit) [inline package.json]: Yes, correct. Will be removed!

File: templates/partials/head.tpl (hunk 5 above)
Comment (radibit) [inline templates/partials/head.tpl]: Yes, nice catch, I forgot to add it!

File: app.js (hunk 2 above)
Comment (radibit) [inline app.js]: As I understand correctly you want to remove the md5 hash set up and add the scripts to the `rev.json`  as well, right?

File: app.js (hunk 1 above)
Comment (radibit) [inline app.js]: Yes, my bad :+1:

File: app.js (hunk 2 above)
Comment (stefanjudis) [inline app.js]: Yap - so that we'll go the same approach with this the loaded assets.

File: templates/partials/head.tpl (hunk 5 above)
Comment (radibit) [inline templates/partials/head.tpl]: Fixed!

File: package.json (hunk 4 above)
Comment (radibit) [inline package.json]: Fixed!

File: app.js (hunk 1 above)
Comment (radibit) [inline app.js]: Fixed!

**Comment by stefanjudis** [comment]
@radibit 
> Also why do we need rev.json in root and in public?

**Comment by radibit** [comment]
@stefanjudis 

The reason for the rev.json in the root is because I'm using this tool [https://www.npmjs.com/package/gulp-json-editor](https://www.npmjs.com/package/gulp-json-editor)

and create the hash during the gulp process. Of course I can just use the file from the root and not copy it to the public.

File: gulpfile.js (hunk 3 above)
Comment (radibit) [inline gulpfile.js]: Can you please give me a more detailed example how that can be achieved?

Comment (stefanjudis) [inline gulpfile.js]: Just put a return before the stream -> `return gulp.src( files.rev )`
and check if it's still working. :)

I had some weird problems with that once and you see all the tasks return the given stream. :)

**Comment by stefanjudis** [comment]
@radibit 

It's not used by the clients out there so I see no reason to have it in public?

Please correct me if I'm wrong. :)

**Comment by radibit** [comment]
@stefanjudis 

You're right. I'm updating this right now.

File: gulpfile.js (hunk 3 above)
Comment (radibit) [inline gulpfile.js]: Updated and working as expected.

**Comment by radibit** [comment]
@stefanjudis @marcobiedermann 

Yesterday I did the minor changes requested by the CR and implemented the style hashing method to the tooling scripts. 

Can you please verify and test the solution?

**Comment by stefanjudis** [comment]
@radibit 

The SVG solution just landed in master. Can you please implement your solution for the svg's? :blush: 

After that I'll make a final check and we're ready to push it out the next release. :tada:

**Comment by radibit** [comment]
@stefanjudis 

The hashing solution using `gulp-rev` is applied for the SVG icons file.
Can you please verify if everything is ok and confirm?

Thanks :smiley:

File: app.js [hunk 6]
```diff
@@ -506,9 +522,21 @@ config.listPages.forEach( function( page ) {
 pages.index = renderPage( 'index' );
 
 app.get( '/', function( req, res ) {
-  res.send( pages.index );
+  if ( req.cookies.maincss ) {
```
Comment (stefanjudis) [inline app.js]: Maybe I'm wrong, but this needs to check for correct revision or?

File: package.json [hunk 7]
```diff
@@ -18,24 +18,32 @@
...
     "MD5": "^1.2.1",
     "async": "^0.9.0",
     "compression": "^1.1.0",
+    "cookie-parser": "^1.3.4",
     "express": "^4.9.0",
     "gulp": "~3.8.0",
     "gulp-autoprefixer": "~2.1.0",
     "gulp-combine-media-queries": "^0.2.0",
     "gulp-concat": "~2.4.1",
+    "gulp-connect": "^2.2.0",
```
Comment (stefanjudis) [inline package.json]: why is this needed?

File: app.js [hunk 8]
```diff
@@ -498,7 +503,18 @@ config.listPages.forEach( function( page ) {
         )
       );
     } else {
-      res.send( pages[ page ] );
+      if ( req.cookies.maincss ) {
+        res.send(
```
Comment (stefanjudis) [inline app.js]: Maybe I'm wrong, but this needs to check for correct revision or?

File: package.json (hunk 7 above)
Comment (radibit) [inline package.json]: I think it was left by mistake. I think it was part of my try to set up a live reload configuration and it's not relevant now, because you already configured the nodemon.

**Comment by stefanjudis** [comment]
@radibit 

also these might not work anymore.

`<use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="/icons.svg?e2b37d1a#icon-arrow"></use>`

**Comment by stefanjudis** [comment]
I have definitely to tackle #114 soonish. :blush:

**Comment by stefanjudis** [comment]
Come on - let's do it. Looking good for me. :)

**Comment by stefanjudis** [comment]
:+1:

**Comment by radibit** [comment]
Thank you :tada: :relaxed:
